### DISCUSSION TITLE
Audio cues: issue tracking

### DISCUSSION THREAD

**Discussion by naorunaoru** [tetele/onju-voice-satellite#69] [Ideas]
As a user, I'd like my smart whoopie cushion to respond with audio feedback. Example cases:
- detecting a wake word (micro or otherwise)
- detecting end of speech
- failing to perform a request
- changing volume

There's a feature request in ESPHome: https://github.com/esphome/feature-requests/issues/2490 which didn't get much attention for some reason. 

There's also a HomeAssistant discussion with some hacks: https://community.home-assistant.io/t/play-sound-when-wakeword-detected/653928/32

Apart from the ESPHome limitation, is there anything specific to Onju Voice that might prevent us from implementing this in the future? From the top of my head I can think of full duplex audio with echo cancellation, as I'm not familiar with ESP-IDF audio pipeline abilities. The custom component used in microwakeword version of this config apparently supports simultaneous listening and playback but it comes with its own tradeoffs. The other issue is that apparently ESPHome can't play multiple audio streams at the same time, which is needed for audio feedback for adjusting the volume while there's music playing. 

The purpose of opening this thread is to share any links related to this feature and discuss possible workarounds. 

I also noticed that people generally suggest playing a prerecorded audio, which isn't quite well suited to ESP32-based systems due to flash size limitations. I'd suggest using a tone generator instead. For example, a sine wave synthesizer with ability to define ADSR volume envelope wouldn't take much EEPROM space and CPU cycles but would allow for instant audio cues without relying on external audio sources. What do you think?

**Comment by tetele** [comment]
> I also noticed that people generally suggest playing a prerecorded audio, which isn't quite well suited to ESP32-based systems due to flash size limitations.

The Onju has 16MB of NOR flash on board which could be used to store such small audio files, but the limitation is due to the fact that ESPHome does not play "files" which are not embedded in the firmware. But the truth is I am nearly clueless about partitions and how to extend the onboard flash with the external one.

**Reply by naorunaoru** [reply]
I did a quick research and it seems like ESPHome doesn't really provide a way to utilize SPI flash, which is sad. Furthermore, I have a feeling that ESPHome itself isn't a very good platform for implementing a voice assistant or really anything that has to do with audio streaming.

A radical approach would be exposing the speaker using Wyoming protocol (or whatever HASS uses for a voice satellite) server running on bare ESP32, but that's way outside of the scope of this repo.

**Reply by naorunaoru** [reply]
Ok, I did a bit more screwing around and found out that we indeed can (in theory) extend the config with an unhealthy dose of custom components. Seems like it's in fact doable and I was a bit critical of ESPHome previously. The only thing I haven't figured out yet is how to put the audio files into the flash without taking the speaker apart.

**Reply by jhbruhn** [reply]
How much space do the audio files actually need? Using the external file component, at least one sound seems to fit the space of the  ESP32-S3 of the S3-Box. That might have different internal flash specs though: https://github.com/esphome/firmware/blob/a7db5f3dde6be30ebfd0ad2543a772a546433a45/wake-word-voice-assistant/esp32-s3-box-3.yaml#L713

**Reply by tetele** [reply]
The Box 3 has an ESP32-N16R8 with 16MB of internal flash. I am honestly not sure if external NOR flash can be used as storage space in the partitions, but if it can, I don't see why we couldn't embed such files in the actual firmware.

**Reply by jhbruhn** [reply]
I just had another look at the ESP32-S3 Datasheet, Onju Schematics and the ESPHome configuration. At least according to the schematics, there should be a ESP32-S3R8 built into the Onju (also confirmed by my PCBWay BOM). That has no internal Flash and 8MB internal PSRAM. 

The flash on the Onju Voice PCB, which is the flash that already contains the ESPHome firmware(!), is a W25Q128-SPI-Flash, which should have 16MByte of Memory. The Onju Voice ESPHome config however, [specifies the `esp32-s3-devkitc-1` board](https://github.com/tetele/onju-voice-satellite/blob/cfd461f730eaa3e749a356930fafbeafa09a04f3/esphome/onju-voice-microwakeword.yaml#L58), which from its platformio config has only 8MByte of Flash [specified](https://docs.platformio.org/en/latest/boards/espressif32/esp32-s3-devkitc-1.html#espressif-esp32-s3-devkitc-1-n8-8-mb-qd-no-psram). That means that we should be able to "double" the available flash memory by correcting the board definition by specifying `flash_size: 16MB` in the esphome config.

Update: Yep that seems to work. Looks like plenty of space for some internal sound files (6MB to be precise).
```
RAM:   [=         ]  10.6% (used 34696 bytes from 327680 bytes)
Flash: [==        ]  17.0% (used 1378649 bytes from 8126464 bytes)
```

Update 2: I think I was wrong about esphome using the flash-size definition from the platform.io board. According to the ESPHome Documentation, it seems to have default to 4MB.

**Reply by dreimer1986** [reply]
Just as a lil stupid idea. The whole Onju is useless without Home Assistant anyways, so why don't we just host these lil audio files on there and stream them? This would make replacing or editing way easier, too.

**Reply by jhbruhn** [reply]
While you are right about the Home Assistant dependency, I don't feel right doing this setup with this Onju Firmware, as that is quite a complicated step for everyone to configure manually, compared to the current state of the firmware. Additionally, I fear that the latency might be high compared to local sounds, especially for a post-wakeword-sound.

I want to make an alternative onju firmware variant this weekend which does not use the media_player but only a speaker component. That should then be able to play local sounds.

**Reply by tetele** [reply]
Apart from not being able to play media on a `speaker`, the biggest loss is volume control. And it's quite loud at the default 100%.

You could maybe go around that by controlling the DAC directly (i haven't researched how to do that).

I do agree, however, that it's tedious to set up URLs pointing to HA to play sounds.

**Reply by jhbruhn** [reply]
Maybe we'll need to implement a custom component for the Onjus DAC based on [this](https://github.com/esphome/voice-kit), as soon as that implementation has matured a bit. But the MAX98357 does not have a digital interface (besides I2S), so all volume setting would be done on the ESP32 Audio side anyways.

**Reply by dreimer1986** [reply]
Well, I added timer support and kept media_player as it is. It was one of the main reasons for me to be able to run Music Assistant over it and thus speaker is out of question for me. And yes I get my timer finished signal off a online source, just like the official ESPHome ones do, too.

Like: https://github.com/esphome/firmware/blob/main/wake-word-voice-assistant/esp32-s3-box-3.yaml
Check the file: setting.

My sorry experiments:
https://github.com/dreimer1986/onju-voice-satellite/blob/main/esphome/onju-voice-microwakeword.yaml

While we are at it... I try to get microwakeword and Assistant itself to have sort of better hearing. To make it wake up I have to sit more than just next to the Onju Device. I tried a few settings you can find under microphone and voice_assistant commented out, but it seems like my current tinkering did not really help here...

P.S. I am more than open for any suggestions regarding my below basic level tinkering. Of course I wait for tetele to have some spare time for continuing his great work, but until now... I thought I could hack a bit myself ^^

**Reply by jhbruhn** [reply]
I think you solved it in the best way possible with keeping media_player right now. But the S3 Box firmware does it a bit differently: It embeds the file during compilation (yes, from an online source) into the firmware image. During runtime, the sound is loaded from the firmware image, so the S3 Box is not doing an online request.

**Reply by dreimer1986** [reply]
I just realized that, too. Sry for the false claim... Aaaanyways. It works and I love the fact that I have one thing less I still had to rely on my Echo device ^^. The list shrinks more and more... If this thing would just listen to commands and wake word a bit better...
